irodsConnector/session.py

```diff
@@ -9,13 +9,14 @@
 import irods.password_obfuscation
 import irods.session
 
-from . import keywords as kw
 import utils
+from . import keywords as kw
 
 
 class Session(object):
     """Irods session operations """
-    _session = None
+    _irods_session = None
+    # Singleton instance initially configured in iBridges.py
     context = utils.context.Context()
```

Comment:
With that we can still create circular properties, we somehow need to break the circle:
```
context.irods_connector.session.context.irods_connector.session.context.irods_connector...
```
I see two options here:
Let only the manager use the context object and pass parameters for the session through, or do not add the manager object to the context object.

Reply:
A person goes to the doctor and says, "Doctor, it hurts when I do this."  The doctor responds, "Then don't do that!"

Does this circular reference cause any issues?  Could it lead to issues?  Garbage collection aside, I can't think of any.

For this case, the manager _cannot_ use the context without potentially having the same issue.  Only `resource` and `session` use the configurations.  So the only possibility I see would be to pass those through the manager.  Not a problem, but it feels to me like a step back to be passing around what can easily be obtained via a singleton.

I would be concerned that future functionality creates a need for more dynamic access to the configurations and we would need to revert to the current situation, but I will see if there are currently any negative effects of making this change.  If I don't have any issues I'll commit it.

Reply:
Unfortunately, the session creation involving the _auth file_ that enables cached password logins requires the name of the _environment file_.  This value is normally passed via the context object.  Also, the recent change to allow using the previously cached environment file as the 'last_ienv' value is also affected by this.  That code needs to update the value in the context object.

I can pass this value too (I was afraid it wouldn't be so simple), but the 'last_ienv' clause will need to be removed if the Session class no longer has access to the context object.

What would you prefer?  I'm suspecting it will be dropping the 'last_ienv' convenience to remove the circular reference.

Reply:
I think we can drop the 'last_ienv' (in fact I have already done so in PR #89 ). It is a nice convenience to set a previous environment, but we can leave that to the front end, where we actually already do that. For the backend I do not see any need, we can throw an error of not enough information was given to instantiate the session.

Reply:
Unfortunately it will not work without far more setup.  Right now, an interactive session can be started with:

```python
import utils
import irodsConnector
context = utils.context.Context()
context.irods_connector = irodsConnector.manager.IrodsConnector()
```

optionally specifying the `irods_env_file` and possibly a password.  For the resource and session objects, if context is not used to pass the configurations and `irods_env_file` value (which now _must_ be manually set), they will all have to be set manually:

```python
import utils
import irodsConnector
context = utils.context.Context()
context.irods_connector = irodsConnector.manager.IrodsConnector()
context.irods_env_file = '~/.irods/irods_environment.json'
context.irods_connector.resource.conf = self.context.ibridges_configuration.config
context.irods_connector.resource.ienv = self.context.irods_environment.config
context.irods_connector.session.conf = self.context.ibridges_configuration.config
context.irods_connector.session.ienv = self.context.irods_environment.config
context.irods_connector.session.irods_env_file = context.irods_env_file
```

This is because the iBridges session and resource objects are already instantiated _before_ the `irods_env_file` value and configuration dictionaries can be passed.  They must be updated.  The iBridges configuration is not as problematic, but must be set the same way.  For the GUI, the `context.irods_env_file` is already set, but the appropriate version of the last 5 lines will need to follow that in the code.

All in all, not so nice!

Reply:
I finally managed the correct method of _passing_ the configurations to the `irodsConnector` classes _after_ instantiation and _prior_ to the iRODS **session** creation in order to remove those pesky circular references.  It appears a little _under-the-table_, but I was already using this method for updating other aspects of the `Context` during a **Close session**.  It works quite well and I hope clear enough.  Testing will show!

The new logging has some issues that i still need figure out and I have not yet tested _interactive_ instantiation of the `Context` and iRODS session.  @chStaiger I'm sure you will let me know where you find problems.